[pytree] add APIs to determine a class is a namedtuple or PyStructSequence

XuehaiPan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #138214
* #113258
* #148569
* __->__ #113257

Changes in this PR:

1. Add `is_structseq` and `is_structseq_class` functions to determine a object or a class is PyStructSequence.
2. Add a generic class `structseq` which can be used as the registration key for PyStructSequence types like `namedtuple` for Named Tuple types.
3. Change `is_namedtuple` to accept subclasses of namedtuple to be namedtuple. Before this PR, only namedtuple class directly created by `collections.namedtuple` or `typing.NamedTuple` were namedtuple classes while their subclasses were not. This PR makes `is_namedtuple` return true for subclasses of namedtuple class.

Resolves #75982. New tests are included in this PR.

- #75982


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @rec

XuehaiPan committed:
[pytree] add APIs to determine a class is a namedtuple or PyStructSequence

[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] add APIs to determine a class is a namedtuple or PyStructSequence"


Resolves #75982

- #75982


[ghstack-poisoned]

XuehaiPan commented on torch/autograd/forward_ad.py:
```diff
@@ -118,16 +117,32 @@ def make_dual(tensor, tangent, *, level=None):
-
-
-class UnpackedDualTensor(_UnpackedDualTensor):
+class UnpackedDualTensor(tuple):
```
Now a subclass of a namedtuple class is also considered as a namedtuple class. So it will be treated as a pytree node. Here we make `UnpackedDualTensor` a custom class (do not use `namedtuple`) to let it still be a leaf type.

albanD replied:
What is the user-visible change related to this? How should we help user understand what we're changing here so they can adapt their code as needed.

XuehaiPan replied:
Users will not need to change their code. The new implementation still looks like a namedtuple and still remains a leaf type in our pytree utilities.

```python
# Create a new instance like before
unpacked = UnpackedDualTensor(primal, tangent)

# Tuple unpacking
primal, tangent = unpacked

# Named access
primal = unpacked.primal
tangent = unpacked.tangent
```

albanD reviewed (commented):
Is this going to change the behavior for every namedtuple class in the wild? That sounds pretty BC-breaking?

XuehaiPan commented:
> Is this going to change the behavior for every namedtuple class in the wild? That sounds pretty BC-breaking?

It will not affect namedtuple classes directly created by `colllections.namedtuple` or `typing.NamedTuple`:

```python
# Not affected by this PR
DirectNamedTuple1 = collections.namedtuple('DirectNamedTuple1', ['field1', 'field2', 'field3'])


# Not affected by this PR
class DirectNamedTuple2(typing.NamedTuple):
    fieldA: int
    fieldB: str
    fieldC: torch.Tensor
```

This PR affects subclasses of the namedtuple classes. They are kind of `namedtuple` but not created directly by `ollections.namedtuple`.

```python
# Not affected by this PR
DirectNamedTuple1 = collections.namedtuple('DirectNamedTuple1', ['field1', 'field2', 'field3'])


# Before this PR. The following type is not a namedtuple. It is a leaf type in pytree.
# After this PR. The following type is a namedtuple. It is a node type in pytree.
class ChildNamedTuple1(DirectNamedTuple1):
    pass


# Not affected by this PR
class DirectNamedTuple2(typing.NamedTuple):
    fieldA: int
    fieldB: str
    fieldC: torch.Tensor


# Before this PR. The following type is not a namedtuple. It is a leaf type in pytree.
# After this PR. The following type is a namedtuple. It is a node type in pytree.
class ChildNamedTuple2(DirectNamedTuple2):
    pass
```

Before this PR:

```python
>>> is_namedtuple_class(DirectNamedTuple1)
True
>>> is_namedtuple_class(DirectNamedTuple2)
True
>>> is_namedtuple_class(ChildNamedTuple1)
False
>>> is_namedtuple_class(ChildNamedTuple2)
False
```

After this PR:

```python
>>> is_namedtuple_class(DirectNamedTuple1)
True
>>> is_namedtuple_class(DirectNamedTuple2)
True
>>> is_namedtuple_class(ChildNamedTuple1)
True
>>> is_namedtuple_class(ChildNamedTuple2)
True
```

------

Some references: a subclass of a dataclass is still a dataclass.

```python
In [1]: import dataclasses

In [2]: @dataclasses.dataclass
   ...: class MyClass:
   ...:     a: int
   ...:     b: str
   ...:     

In [3]: dataclasses.is_dataclass(MyClass)
Out[3]: True

In [4]: class OtherClass(MyClass):
   ...:     pass
   ...:     

In [5]: dataclasses.is_dataclass(OtherClass)
Out[5]: True
```

XuehaiPan committed:
Update on "[pytree] add APIs to determine a class is a namedtuple or PyStructSequence"


Resolves #75982

- #75982


[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] add APIs to determine a class is a namedtuple or PyStructSequence"


Resolves #75982

- #75982


[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] add APIs to determine a class is a namedtuple or PyStructSequence"


Resolves #75982. New tests are included in this PR.

- #75982


[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] add APIs to determine a class is a namedtuple or PyStructSequence"


Resolves #75982. New tests are included in this PR.

- #75982


[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] add APIs to determine a class is a namedtuple or PyStructSequence"


Resolves #75982. New tests are included in this PR.

- #75982


[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] add APIs to determine a class is a namedtuple or PyStructSequence"


Resolves #75982. New tests are included in this PR.

- #75982


[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] add APIs to determine a class is a namedtuple or PyStructSequence"


Resolves #75982. New tests are included in this PR.

- #75982


[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] add APIs to determine a class is a namedtuple or PyStructSequence"


Resolves #75982. New tests are included in this PR.

- #75982


[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] add APIs to determine a class is a namedtuple or PyStructSequence"


Resolves #75982. New tests are included in this PR.

- #75982


[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] add APIs to determine a class is a namedtuple or PyStructSequence"


Resolves #75982. New tests are included in this PR.

- #75982


[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] add APIs to determine a class is a namedtuple or PyStructSequence"


Resolves #75982. New tests are included in this PR.

- #75982


[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] add APIs to determine a class is a namedtuple or PyStructSequence"


Resolves #75982. New tests are included in this PR.

- #75982


[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] add APIs to determine a class is a namedtuple or PyStructSequence"


Resolves #75982. New tests are included in this PR.

- #75982


[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] add APIs to determine a class is a namedtuple or PyStructSequence"


Resolves #75982. New tests are included in this PR.

- #75982


[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] add APIs to determine a class is a namedtuple or PyStructSequence"


Resolves #75982. New tests are included in this PR.

- #75982


[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] add APIs to determine a class is a namedtuple or PyStructSequence"


Resolves #75982. New tests are included in this PR.

- #75982


[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] add APIs to determine a class is a namedtuple or PyStructSequence"


Resolves #75982. New tests are included in this PR.

- #75982


[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] add APIs to determine a class is a namedtuple or PyStructSequence"


Resolves #75982. New tests are included in this PR.

- #75982


[ghstack-poisoned]

zou3519 commented on torch/utils/_pytree.py:
```diff
@@ -282,6 +291,78 @@ def _private_register_pytree_node(
+
+
+# Reference: https://github.com/metaopt/optree/blob/main/optree/typing.py
+def is_namedtuple_class(cls: type) -> bool:
+    """Return whether the class is a subclass of namedtuple."""
```
This PR also makes subclasses of namedtuple into pytree nodes. What's the behavior in JAX?

zou3519 replied:
Thanks for clarifying.

zou3519 reviewed (commented):
The changes to PyTree seem fine to me. This is technically BC-breaking, but it aligns us more with JAX's pytree and I don't expect anyone out in the wild to depend on this behavior.

I'm a little confused at what is going on with UnpackedDualTensor. What happens if we do allow pytree to peek into it? Does anything in the test suite break?

XuehaiPan committed:
Update on "[pytree] add APIs to determine a class is a namedtuple or PyStructSequence"


Changes in this PR:

1. Add `is_structseq` and `is_structseq_class` functions to determine a object or a class is PyStructSequence.
2. Change `is_namedtuple` to accept subclasses of namedtuple to be namedtuple. Before this PR, only namedtuple class directly created by `collections.namedtuple` or `typing.NamedTuple` were namedtuple classes while their subclasses were not. This PR makes `is_namedtuple` return true for subclasses of namedtuple class.

Resolves #75982. New tests are included in this PR.

- #75982


[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] add APIs to determine a class is a namedtuple or PyStructSequence"


Changes in this PR:

1. Add `is_structseq` and `is_structseq_class` functions to determine a object or a class is PyStructSequence.
2. Change `is_namedtuple` to accept subclasses of namedtuple to be namedtuple. Before this PR, only namedtuple class directly created by `collections.namedtuple` or `typing.NamedTuple` were namedtuple classes while their subclasses were not. This PR makes `is_namedtuple` return true for subclasses of namedtuple class.

Resolves #75982. New tests are included in this PR.

- #75982


[ghstack-poisoned]

zou3519 commented:
I slept on this a little more. I'm concerned about the BC for namedtuple subclasses. They are definitely more common than structseq (and we have one in our codebase with the UnpackedDualTensor). I would punt on this PR for now and build the correct behavior into our future torch.utils.pytree API. WDYT?

XuehaiPan commented:
> I'm concerned about the BC for namedtuple subclasses. They are definitely more common than structseq (and we have one in our codebase with the UnpackedDualTensor).

@zou3519 The JAX pytree flattens namedtuple types and their subclasses as pytree nodes.

The `UnpackedDualTensor` became a leaf node from in:

- #76492

Before that, `UnpackedDualTensor` was a pytree node class which directly created by `collections.namedtuple()`.

Since we now have tests that require the `UnpackedDualTensor` instances to be leaf nodes in:

https://github.com/pytorch/pytorch/blob/a6736ac8518aff7bb88aa0454ba0de43e49a57b3/torch/testing/_internal/composite_compliance.py#L542-L544

https://github.com/pytorch/pytorch/blob/a6736ac8518aff7bb88aa0454ba0de43e49a57b3/torch/testing/_internal/composite_compliance.py#L576-L578

I leave the `UnpackedDualTensor` class as a leaf node by inheriting from `tuple` directly.

zou3519 commented:
To clarify, I'm concerned that we previously were not treating subclasses of namedtuple classes as pytrees, but this PR changes it so that we are treating them as pytrees. This will end up silently changing the behavior for any users who have subclasses of namedtuples. Is my understanding of the PR correct?

XuehaiPan commented:
> To clarify, I'm concerned that we previously were not treating subclasses of namedtuple classes as pytrees, but this PR changes it so that we are treating them as pytrees. This will end up silently changing the behavior for any users who have subclasses of namedtuples. Is my understanding of the PR correct?

Yes.

> any users who have subclasses of namedtuples

I don't think this is a common practice in the wild. Tracing back of the change history of `UnpackedDualTensor`, it was created directly with `namedtuple()`. Also, before this PR, the `UnpackedDualTensor` inherits `_UnpackedDualTensor` with an empty class body.

XuehaiPan committed:
Update on "[pytree] add APIs to determine a class is a namedtuple or PyStructSequence"


Changes in this PR:

1. Add `is_structseq` and `is_structseq_class` functions to determine a object or a class is PyStructSequence.
2. Change `is_namedtuple` to accept subclasses of namedtuple to be namedtuple. Before this PR, only namedtuple class directly created by `collections.namedtuple` or `typing.NamedTuple` were namedtuple classes while their subclasses were not. This PR makes `is_namedtuple` return true for subclasses of namedtuple class.

Resolves #75982. New tests are included in this PR.

- #75982


[ghstack-poisoned]

XuehaiPan committed:
Update on "[pytree] add APIs to determine a class is a namedtuple or PyStructSequence"


Changes in this PR:

1. Add `is_structseq` and `is_structseq_class` functions to determine a object or a class is PyStructSequence.
2. Change `is_namedtuple` to accept subclasses of namedtuple to be namedtuple. Before this PR, only namedtuple class directly created by `collections.namedtuple` or `typing.NamedTuple` were namedtuple classes while their subclasses were not. This PR makes `is_namedtuple` return true for subclasses of namedtuple class.

Resolves #75982. New tests are included in this PR.

- #75982


[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

zou3519 reviewed (changes requested):
tests are failing

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan commented on torch/autograd/forward_ad.py:
```diff
@@ -129,16 +128,22 @@ def make_dual(tensor, tangent, *, level=None):
-
-
-class UnpackedDualTensor(_UnpackedDualTensor):
+# NOTE: It is considered a opaque leaf in pytree utilities.
```
Removed the `__torch_dynamo_can_constant_fold_through__` part.

XuehaiPan commented on benchmarks/dynamo/common.py:
```diff
@@ -1413,7 +1413,7 @@ def load(cls, model, example_inputs):
                 example_outputs = copy.deepcopy(model)(*example_args, **example_kwargs)
 
-            if pytree._is_namedtuple_instance(example_outputs):
+            if pytree.is_namedtuple_instance(example_outputs):
```
I think we should provide a standard `is_namedtuple` API in public. There are three `is_namedtuple` functions in the codebase. We can merge them in the future.

https://github.com/pytorch/pytorch/blob/346bbefa630b58fda5453373e2a3bdcc32236a16/torch/utils/_pytree.py#L778-L787

https://github.com/pytorch/pytorch/blob/346bbefa630b58fda5453373e2a3bdcc32236a16/torch/nn/parallel/scatter_gather.py#L13-L26

https://github.com/pytorch/pytorch/blob/346bbefa630b58fda5453373e2a3bdcc32236a16/torch/_dynamo/utils.py#L2074-L2106

zou3519 replied:
okey doke

XuehaiPan commented on torch/autograd/forward_ad.py:
```diff
@@ -129,16 +128,22 @@ def make_dual(tensor, tangent, *, level=None):
+# NOTE: It is considered a opaque leaf in pytree utilities.
+#       We need to make this a quasi-namedtuple type that is not a pytree node while still keep
+#       Dynamo trace this type as a namedtuple type.
+class UnpackedDualTensor(NamedTuple):
```
I add a more detailed comment in the code about this.

XuehaiPan commented on torch/utils/_cxx_pytree.py:
```diff
@@ -234,6 +240,36 @@ def _is_pytreespec_instance(obj: Any, /) -> TypeIs[TreeSpec]:
     return isinstance(obj, TreeSpec)
 
 
+def is_namedtuple(obj: Union[object, type]) -> bool:
+    """Return whether the object is an instance of namedtuple or a subclass of namedtuple."""
+    return optree.is_namedtuple(obj)
+
+
+def is_namedtuple_class(cls: type) -> bool:
+    """Return whether the class is a subclass of namedtuple."""
+    return optree.is_namedtuple_class(cls)
+
+
+def is_namedtuple_instance(obj: object) -> bool:
+    """Return whether the object is an instance of namedtuple."""
+    return optree.is_namedtuple_instance(obj)
+
+
+def is_structseq(obj: Union[object, type]) -> bool:
+    """Return whether the object is an instance of PyStructSequence or a class of PyStructSequence."""
+    return optree.is_structseq(obj)
+
+
+def is_structseq_class(cls: type) -> bool:
+    """Return whether the class is a class of PyStructSequence."""
+    return optree.is_structseq_class(cls)
+
+
+def is_structseq_instance(obj: object) -> bool:
+    """Return whether the object is an instance of PyStructSequence."""
+    return optree.is_structseq_instance(obj)
```
Now the implementations of `is_{namedtuple,structseq}{,_class,_instance}` for both pytree impls are aligned while the C++ pytree contains a caching mechanism that runs faster.

I can import the implementation from the Python pytree if you think that ones are better.

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

zou3519 commented on torch/autograd/forward_ad.py:
```diff
@@ -129,16 +128,33 @@ def make_dual(tensor, tangent, *, level=None):
+    tangent: Optional[torch.Tensor]
+
+
+# remove `_asdict` method to make it an opaque leaf (pytree checks `_fields`, `_make`, `_asdict`)
+del UnpackedDualTensor._asdict
```
@soulitzer you okay with this? I'm fine with it

zou3519 reviewed (commented):
Looks good to me. Just want confirmation from @soulitzer on the forward_ad change

zou3519 commented on torch/autograd/forward_ad.py:
```diff
@@ -129,16 +128,33 @@ def make_dual(tensor, tangent, *, level=None):
+# To keep backward-compatibility and make the unit tests happy, we need to make this a quasi-namedtuple
+# type that is not a pytree node while still keep Dynamo trace this type as a namedtuple type.
+# We achieve this by removing the `_asdict` attribute. See the comment below the class definition.
+class UnpackedDualTensor(NamedTuple):
```
@XuehaiPan, @soulitzer and I discussed this a bit. What exactly is the BC issue that you ran into? It's possible that it doesn't matter.

Dynamo doesn't actually support forward_ad (but it does support functorch.jvp). In the functorch.jvp case the user should never actually see an UnpackedDualTensor.

XuehaiPan replied:
> What exactly is the BC issue that you ran into?

The BC issue is that this class was not previously a pytree node. Some tests rely on this behavior. See https://github.com/pytorch/pytorch/pull/113257#issuecomment-1850365365

https://github.com/pytorch/pytorch/blob/915b9c80ab59a1dfc148cc476deaef926b2624b5/torch/testing/_internal/composite_compliance.py#L554-L556

If we want to remove the workaround in this PR, we can refactor the tests with:

```python
expected = tree_map(fwAD.unpack_dual, expected) 
expected_primals = tree_map(lambda x: x.primal, expected, is_leaf=lambda type(x) is UnpackedDualTensor) 
expected_tangents = tree_map(lambda x: x.tangent, expected, is_leaf=lambda type(x) is UnpackedDualTensor) 
```

XuehaiPan replied:
> If that is the only issue then let's refactor the test.

I opened a follow up PR to address this:

- #148151

zou3519 replied:
@XuehaiPan all PRs should be able to stand by themselves. This means this PR should not be deleting _asdict, even if you have a follow-up PR that fixes the issue

XuehaiPan replied:
I squashed the PRs.

XuehaiPan committed:
Update

[ghstack-poisoned]

zou3519 reviewed (changes requested):
Let's not delete _asdict in this PR

XuehaiPan committed:
Update

[ghstack-poisoned]

zou3519 commented on torch/autograd/forward_ad.py:
```diff
@@ -129,16 +128,15 @@ def make_dual(tensor, tangent, *, level=None):
-
-
-class UnpackedDualTensor(_UnpackedDualTensor):
+class UnpackedDualTensor(NamedTuple):
```
@XuehaiPan did you need to subclass NamedTuple? Why did we need to change the original code?

XuehaiPan replied:
We can keep it as it is. I can revert this if that is preferred.

The original code was previously changed here https://github.com/pytorch/pytorch/pull/76492/files#diff-a4efad4ba18fffdfb264c21e5475997a24a743089a899f8ec1a5ff962c6738d9 3 years ago. That was a workaround to add `__doc__` to a namedtuple type created by `collections.namedtuple`. Subclassing `typing.NamedTuple` is a more modern way to create a namedtuple type.

zou3519 replied:
no this is good, thank you for explaiing

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan commented:
@pytorchbot merge

jovianjaison commented:
@pytorchbot revert -m "breaking internal builds" -c ghfirst

XuehaiPan commented:
> "breaking internal builds"

Any clue for this?

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan commented:
@pytorchbot merge

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan commented:
@pytorchbot merge

ZainRizvi commented:
@pytorchbot revert -c ghfirst -m "Sorry but this is breaking internally. @zou3519 can you please help land this internally? See the sigmoid tests in D71198793 for details. To validate the fixes internally, you can follow the instructions here: https://fburl.com/fixing-ghfirst-reverts"

Error looks like: 
```
RuntimeError: Unknown pytree node type: NamedTupleIn
```

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan commented:
Any news on this? cc @zou3519

zou3519 commented:
My best guess is that previously a NamedTuple was being treated as a tuple for the purposes of pytree but it is now being treated as a NamedTuple with this PR

angelayi commented:
The issue is that this PR has a backwards-compatibility breaking change on the treespec. Internally, we have been saving the treespec and loading it for models so it broke some things when the behavior changed.

```python
DirectNamedTuple1 = namedtuple("DirectNamedTuple1", ["x", "y"])
treespec = pytree.tree_flatten(DirectNamedTuple1(1, 2))[1]
print(treespec)
# Previously: TreeSpec(namedtuple, <class '__main__.DirectNamedTuple1'>, [*, *])
# Now: PyTreeSpec(DirectNamedTuple1(x=*, y=*), NoneIsLeaf, namespace='torch')
```

XuehaiPan commented:
> Internally, we have been saving the treespec and loading it for models

@angelayi This PR only changes the behavior during flattening. If the treespec is created and saved by the old code, there should be not break by this PR.

angelayi commented:
Ah you're right sorry, it's not BC breaking, it's FC breaking (but that's not a super strong requirement). The internal failure is that since the type is now the actual namedtuple class instead of "namedtuple", our internal cpp implementation of pytree fails since "we didn't do a pytree registration for that type".

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan commented:
> The internal failure is that since the type is now the actual namedtuple class instead of "namedtuple", our internal cpp implementation of pytree fails since "we didn't do a pytree registration for that type".

Is there anything I can help? The 2.7.0 release branch is finalized today.

angelayi commented:
Sorry, what I meant is that this PR should not change the TreeSpec schema since it's breaking assumptions for frameworks downstream, like our cpp implementation.

XuehaiPan commented:
> it's breaking assumptions for frameworks downstream, like our cpp implementation.

The assumption would be that the type can be reconstructed by the field names, which is not fulfilled for named tuple subclasses. Is it correct?

angelayi commented:
The assumption is that for namedtuple pytrees, the treespec looks like `TreeSpec(namedtuple, <class '__main__.DirectNamedTuple1'>, [*, *])`

XuehaiPan commented:
> The assumption is that for namedtuple pytrees, the treespec looks like `TreeSpec(namedtuple, <class '__main__.DirectNamedTuple1'>, [*, *])`

This PR does not change that.

```python
In [1]: import torch.utils._pytree as pytree

In [2]: from typing import NamedTuple

In [3]: class DirectNamedTuple(NamedTuple):
   ...:     x: int
   ...:     y: int
   ...:     z: int
   ...:     

In [4]: class ChildNamedTuple(DirectNamedTuple):
   ...:     pass
   ...:     

In [5]: pytree.tree_structure(DirectNamedTuple(1, 2, 3))
Out[5]: 
TreeSpec(namedtuple, <class '__main__.DirectNamedTuple'>, [*,
  *,
  *])

In [6]: pytree.tree_structure(ChildNamedTuple(1, 2, 3))
Out[6]: 
TreeSpec(namedtuple, <class '__main__.ChildNamedTuple'>, [*,
  *,
  *])
```

angelayi commented:
```
(Pdb) from collections import namedtuple
(Pdb) import torch.utils._pytree as pytree
(Pdb) DirectNamedTuple1 = namedtuple("DirectNamedTuple1", ["x", "y"])
(Pdb) pytree._register_namedtuple(DirectNamedTuple1, serialized_type_name="DirectNamedTuple1")
(Pdb) pytree.tree_flatten(DirectNamedTuple1(1, 2))[1]
TreeSpec(DirectNamedTuple1, <class 'sigmoid.inference.test.e2e_test_utils.DirectNamedTuple1'>, [*,
  *])
```

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan commented:
> ```
> (Pdb) from collections import namedtuple
> (Pdb) import torch.utils._pytree as pytree
> (Pdb) DirectNamedTuple1 = namedtuple("DirectNamedTuple1", ["x", "y"])
> (Pdb) pytree._register_namedtuple(DirectNamedTuple1, serialized_type_name="DirectNamedTuple1")
> (Pdb) pytree.tree_flatten(DirectNamedTuple1(1, 2))[1]
> TreeSpec(DirectNamedTuple1, <class 'sigmoid.inference.test.e2e_test_utils.DirectNamedTuple1'>, [*,
>   *])
> ```

@angelayi Thanks for the hint. I updated the PR and resolved this in the latest commit.

XuehaiPan committed:
Update

[ghstack-poisoned]

XuehaiPan commented on torch/_export/serde/serialize.py:
```diff
@@ -1243,7 +1243,7 @@ def serialize_treespec(self, treespec):
             if ts.type is None:
                 return
-            if ts.type == namedtuple:
+            if ts.type is namedtuple or pytree.is_namedtuple_class(ts.type):
```
NB: Due to this, previously, all tests passed in OSS while break internally. With the latest commit, this can be reverted.

XuehaiPan commented:
@pytorchbot merge

XuehaiPan commented:
@pytorchbot merge
